tools/bro-to-vast/bro-to-vast.cpp

```diff
@@ -11,6 +11,156 @@
+  std::vector<broker::data> result;
+  // TODO: get actual results from VAST.
+  for (auto i : {42, 43, 44, 45, 46})
+    result.emplace_back(broker::vector{i});
+  return result;
+}
+
+// Parses Broker data as Bro event.
+caf::expected<std::pair<std::string, std::string>>
+parse_query_event(const broker::data& x) {
+  std::pair<std::string, std::string> result;
+  auto event = broker::bro::Event(x);
+  if (event.name() != "VAST::query")
+    return make_error(vast::ec::parse_error, "invalid event name", event.name());
+  if (event.args().size() != 2)
+    return make_error(vast::ec::parse_error, "invalid number of arguments");
+  auto query_id = caf::get_if<std::string>(&event.args()[0]);
+  if (!query_id)
+    return make_error(vast::ec::parse_error, "invalid type of 1st argument");
+  result.first = *query_id;
+  if (!vast::parsers::uuid(*query_id))
+    return make_error(vast::ec::parse_error, "invalid query UUID", *query_id);
+  auto expression = caf::get_if<std::string>(&event.args()[1]);
+  if (!expression)
+    return make_error(vast::ec::parse_error, "invalid type of 2nd argument");
+  if (!vast::parsers::expr(*expression))
+    return make_error(vast::ec::parse_error, "invalid query expression",
+                      *expression);
+  result.second = *expression;
+  return result;
+}
+
+} // namespace <anonymous>
+
 int main(int argc, char** argv) {
-  return 0;
+  // Parse the command line.
+  config cfg;
+  cfg.parse(argc, argv);
+  std::string address = caf::get_or(cfg, "address", default_address);
+  uint16_t port = caf::get_or(cfg, "port", default_port);
+  std::string control_topic = caf::get_or<std::string>(cfg, "control-topic",
+                                                       default_control_topic);
+  std::string data_topic = caf::get_or<std::string>(cfg, "data-topic",
+                                                    default_data_topic);
+  // Create a Broker endpoint.
+  auto endpoint = broker::endpoint{std::move(cfg)};
+  endpoint.listen(address, port);
+  // Subscribe to the control channel.
+  auto subscriber = endpoint.make_subscriber({control_topic});
+  // Block until Bro peers with us.
+  auto receive_statuses = true;
+  auto status_subscriber = endpoint.make_status_subscriber(receive_statuses);
+  auto peered = false;
+  while (!peered) {
+    auto msg = status_subscriber.get();
+    caf::visit(vast::detail::overload(
+      [&](broker::none) {
+        // timeout
+      },
+      [&](broker::error error) {
+        std::cerr << to_string(error) << std::endl;
+      },
+      [&](broker::status status) {
+        if (status == broker::sc::peer_added)
+          peered = true;
+        else
+          std::cerr << to_string(status) << std::endl;
+      }
+    ), msg);
+  };
+  std::cerr << "established peering successfully" << std::endl;
+  // Process queries from Bro.
+  auto done = false;
+  while (!done) {
+    std::cerr << "waiting for commands" << std::endl;
```

Comment:
I've never used `clog`, actually. My mental model contains *standard output* and *standard error*, which also make sense outside of C++. AFAIK `clog` is only a buffered version of `cerr`. Where do you see the benefit here?

Reply:
I see a benefit in the spelling. c**log** suggests logging. But I'm far from adamant about this, and symmetry to the unix io model is fine as well.

Reply:
I'd say let's postpone the discussion of `cerr` vs `clog`. We have other places in the code-base where such refactoring would make sense. Since we've never used `clog` elsewhere, I'd prefer to stick with `cerr` for now and reconsider `clog` for the entire codebase at some later point in time.